File: src/main/java/pcb/uwu/impl/fundamental/Yards.java
package pcb.uwu.impl.fundamental;

import pcb.uwu.amounts.fundamental.Length;
import pcb.uwu.core.BigDecimalAmount;
import pcb.uwu.core.Magnitude;
import pcb.uwu.core.UnitAmount;
import pcb.uwu.units.fundamental.LengthUnit;

import java.math.BigDecimal;
import java.math.MathContext;

import static pcb.uwu.units.fundamental.LengthUnit.YARD;

public class Yards extends Length {

	// region constructors

	public Yards(Number value) {
		super(value, YARD);
	}

	public Yards(Number value, Magnitude magnitude) {
		super(value, magnitude, YARD);
	}

	public Yards(String value) {
		super(value, YARD);
	}

	public Yards(String value, Magnitude magnitude) {
		super(value, magnitude, YARD);
	}

	public Yards(BigDecimal value) {
		super(value, YARD);
	}

	public Yards(BigDecimal value, Magnitude magnitude) {
		super(value, magnitude, YARD);
	}

	public Yards(BigDecimalAmount amount) {
		super(amount, YARD);
	}

	public Yards(BigDecimalAmount amount, Magnitude magnitude) {
		super(amount, magnitude, YARD);
	}

	// endregion

	// region implement UnitAmount

	@Override
	public Yards plus(UnitAmount<LengthUnit> other, MathContext mathContext) {
		return new Yards(plusAmount(other, mathContext));
	}

	@Override
	public Yards minus(UnitAmount<LengthUnit> other, MathContext mathContext) {
		return new Yards(minusAmount(other, mathContext));
	}

	@Override
	public Yards multipliedBy(BigDecimal other, MathContext mathContext) {
		return new Yards(multipliedByScalar(other, mathContext));
	}

	@Override
	public Yards dividedBy(BigDecimal other, MathContext mathContext) {
		return new Yards(dividedByScalar(other, mathContext));
	}

	@Override
	public Yards convertTo(LengthUnit unit) {
		return new Yards(getAmountIn(unit));
	}

	// endregion

}

File: implementation/src/main/java/pcb/uwu/amount/base/Yards.kt
package pcb.uwu.amount.base

import pcb.uwu.core.Magnitude
import pcb.uwu.core.Magnitude.NATURAL
import pcb.uwu.core.UnitAmount
import pcb.uwu.unit.base.LengthUnit
import pcb.uwu.unit.base.YardUnit

class Yards : Length
{
    @JvmOverloads
    constructor(amount: Number,
                magnitude: Magnitude = NATURAL)
            : super(amount, magnitude, YardUnit)

    @JvmOverloads
    constructor(amount: String,
                magnitude: Magnitude = NATURAL)
            : super(amount, magnitude, YardUnit)

    // region UnitAmount

    override fun plus(length: UnitAmount<LengthUnit>) =
        Yards(this.amount + (length to YardUnit).amount)

    override fun minus(length: UnitAmount<LengthUnit>) =
        Yards(this.amount - (length to YardUnit).amount)

    override fun times(scalar: Number) =
        Yards(this.amount * scalar)

    override fun div(scalar: Number) =
        Yards(this.amount / scalar)

    // endregion
}
